### PR TITLE
Enhance IMDMClient for CAD

### PR DESCRIPTION
**Pull request by liadtal**
* Upgrade the version of the ARM and MDM nugets
* Change MDM client interface to be bound to a subscription, not to a specific resource
* Enhance MDM client to get multiple metrics in a single call
* Add method to the MDM client to get metric definitions by resource and service type

### CODE REVIEW

File: src/SDK/SmartDetectorsSDK/Mdm/MdmMetricDefinition.cs [hunk 1]
```diff
@@ -62,5 +62,14 @@ public MdmMetricDefinition(string name, IReadOnlyList<string> dimensions, bool?
...
         /// </summary>
         public Aggregation? PrimaryAggregationType { get; }
+
+        /// <summary>
+        /// Returns the MDM metric definition name.
+        /// </summary>
+        /// <returns>The MDM metric definition name.</returns>
+        public override string ToString()
+        {
+            return this.Name;
```
Comment (aviled) [inline src/SDK/SmartDetectorsSDK/Mdm/MdmMetricDefinition.cs]: >return this.Name; [](start = 12, length = 17)

Seems weird.
Why do you need this?

File: src/SDK/SmartDetectorsPackage/SmartDetectorsPackage.nuproj [hunk 2]
```diff
@@ -22,7 +22,7 @@
   <Import Project="$(NuProjPath)\NuProj.props" Condition="Exists('$(NuProjPath)\NuProj.props')" />
   <PropertyGroup Label="Configuration">
     <Id>Microsoft.Azure.Monitoring.SmartDetectors</Id>
-    <Version>0.2.11.0</Version>
+    <Version>0.2.12.0</Version>
```
Comment (yanivid3) [inline src/SDK/SmartDetectorsPackage/SmartDetectorsPackage.nuproj]: >2 [](start = 15, length = 1)

breaking changes, consider 0.3.0

File: src/SDK/SmartDetectorsSDK/IAnalysisServicesFactory.cs [hunk 3]
```diff
@@ -40,10 +40,10 @@ public interface IAnalysisServicesFactory
         /// <summary>
         /// Creates an instance of <see cref="IMdmClient"/>, used to fetch the resource metrics from MDM.
         /// </summary>
-        /// <param name="resource">The resource to analyze.</param>
+        /// <param name="subscriptionId">The subscription Id</param>
         /// <param name="cancellationToken">The cancellation token</param>
         /// <returns>A <see cref="Task{TResult}"/>, running the current operation, returning the MDM client.</returns>
-        Task<IMdmClient> CreateMdmClientAsync(ResourceIdentifier resource, CancellationToken cancellationToken);
+        Task<IMdmClient> CreateMdmClientAsync(string subscriptionId, CancellationToken cancellationToken);
```
Comment (yanivid3) [inline src/SDK/SmartDetectorsSDK/IAnalysisServicesFactory.cs]: >string [](start = 46, length = 6)

wht not guid?

[the remark this answers is not in the document]
Comment (liadtal) [reply src/SDK/SmartDetectorsSDK/IAnalysisServicesFactory.cs]: Yes, it's a breaking change - but no detector is currently calling this method, so it should be OK to release it even without changing existing detectors.

---
In reply to: [201588774](https://github.com/liadtal/Smart-Alerts/pull/182#discussion_r201588774) [](ancestors = 201588774)

Comment (liadtal) [reply src/SDK/SmartDetectorsSDK/IAnalysisServicesFactory.cs]: We use string everyehere else... See ResourceIdentifier, for example

---
In reply to: [201615322](https://github.com/liadtal/Smart-Alerts/pull/182#discussion_r201615322) [](ancestors = 201615322)

File: src/SDK/SmartDetectorsPackage/SmartDetectorsPackage.nuproj (hunk 2 above)
Comment (liadtal) [reply src/SDK/SmartDetectorsPackage/SmartDetectorsPackage.nuproj]: Not sure - it is a breaking change in a feature (IMdmClient) that nobody used yet...

---
In reply to: [201615069](https://github.com/liadtal/Smart-Alerts/pull/182#discussion_r201615069) [](ancestors = 201615069)

File: src/SDK/SmartDetectorsSDK/Mdm/IMdmClient.cs [hunk 4]
```diff
@@ -38,12 +47,13 @@ public interface IMdmClient
         Task<IEnumerable<MdmQueryResult>> GetResourceMetricsAsync(string resourceFullUri, QueryParameters queryProperties, CancellationToken cancellationToken);
 
         /// <summary>
-        /// Get the resource metric values from MDM, based on the Azure Resource Service (For example: if the resource is a storage account, possible services are BLOB, Queue, and Table)
+        /// Get the resource metric values from MDM, based on the resource and service (for example: if the resource is a storage account, possible services are BLOB, Queue, Table and File)
         /// </summary>
-        /// <param name="azureResourceService">Specific azure resource for which we want to fetch metrics</param>
+        /// <param name="resource">The Azure resource for which we want to fetch metrics</param>
+        /// <param name="azureResourceService">The Azure resource's service type</param>
         /// <param name="queryProperties">MDM properties to be used when fetching data from MDM. All fields are optional</param>
```
Comment (yanivid3) [inline src/SDK/SmartDetectorsSDK/Mdm/IMdmClient.cs]: >MDM [](start = 42, length = 3)

notice to update also the "MDM" in documentations

File: src/SDK/SmartDetectorsShared/Clients/AnalysisServicesFactory.cs [hunk 5]
```diff
@@ -107,16 +107,16 @@ public AnalysisServicesFactory(ITracer tracer, IHttpClientWrapper httpClientWrap
         /// <summary>
         /// Creates an instance of <see cref="IMdmClient"/>, used to fetch the resource metrics from MDM.
```
Comment (yanivid3) [inline src/SDK/SmartDetectorsShared/Clients/AnalysisServicesFactory.cs]: >resource [](start = 79, length = 8)

subscription resource

File: src/SDK/SmartDetectorsShared/Clients/MdmClient.cs [hunk 6]
```diff
@@ -95,7 +93,20 @@ public MdmClient(ITracer tracer, ResourceIdentifier resourceIdentifier, IMonitor
...
+        }
+
+        /// <summary>
+        /// Get the resource metric definitions from MDM
+        /// </summary>
+        /// <param name="resource">The Azure resource for which we want to fetch metric definitions</param>
+        /// <param name="azureResourceService">The Azure resource's service type</param>
+        /// <param name="cancellationToken">Cancellation Token for the async operation</param>
+        /// <returns>A <see cref="Task{TResult}"/> object that represents the asynchronous operation, returning the list of metric definitions</returns>
+        public async Task<IEnumerable<MdmMetricDefinition>> GetResourceMetricDefinitionsAsync(ResourceIdentifier resource, ServiceType azureResourceService, CancellationToken cancellationToken)
```
Comment (yanivid3) [inline src/SDK/SmartDetectorsShared/Clients/MdmClient.cs]: >GetResourceMetricDefinitionsAsync [](start = 60, length = 33)

in which scenario do we use the "GetDefinitions"? I don't remember why we kept it... it is needed only for investigation phase, and there we can use PowerShell

File: src/SDK/SmartDetectorsPackage/SmartDetectorsPackage.nuproj (hunk 2 above)
Comment (liadtal) [reply src/SDK/SmartDetectorsPackage/SmartDetectorsPackage.nuproj]: On second thought - you're right, changed to 0.3.0.0

---
In reply to: [201619043](https://github.com/liadtal/Smart-Alerts/pull/182#discussion_r201619043) [](ancestors = 201619043,201615069)

File: src/SDK/SmartDetectorsShared/Clients/MdmClient.cs (hunk 6 above)
Comment (liadtal) [reply src/SDK/SmartDetectorsShared/Clients/MdmClient.cs]: I added it for CAD

---
In reply to: [201634088](https://github.com/liadtal/Smart-Alerts/pull/182#discussion_r201634088) [](ancestors = 201634088)

File: src/SDK/SmartDetectorsShared/Clients/AnalysisServicesFactory.cs (hunk 5 above)
Comment (liadtal) [reply src/SDK/SmartDetectorsShared/Clients/AnalysisServicesFactory.cs]: I don't think it makes the comment clearer...

---
In reply to: [201632436](https://github.com/liadtal/Smart-Alerts/pull/182#discussion_r201632436) [](ancestors = 201632436)

**Pull request merged**